library/core/src/sync/atomic.rs

```diff
@@ -24,25 +24,43 @@
-//! an `atomic_ref` in C++; the `atomic_ref` is destroyed when the lifetime of the shared reference
-//! ends. A Rust atomic type that is exclusively owned or behind a mutable reference does *not*
-//! correspond to an “atomic object” in C++, since the underlying primitive can be mutably accessed,
-//! for example with `get_mut`, to perform non-atomic operations.
+//! Rust atomics currently follow the same rules as [C++20 atomics][cpp], specifically the rules
+//! from the [`intro.races`][cpp-intro.races] section, without the "consume" memory ordering. Since
+//! C++ uses an object-based memory model whereas Rust is access-based, a bit of translation work
+//! has to be done to apply the C++ rules to Rust: whenever C++ talks about "the value of an
+//! object", we understand that to mean the resulting bytes obtained when doing a read. When the C++
+//! standard talks about "the value of an atomic object", this refers to the result of doing an
+//! atomic load (via the operations provided in this module). A "modification of an atomic object"
+//! refers to an atomic store.
 //!
-//! [cpp]: https://en.cppreference.com/w/cpp/atomic
+//! The end result is *almost* equivalent to saying that creating a *shared reference* to one of the
+//! Rust atomic types corresponds to creating an `atomic_ref` in C++, with the `atomic_ref` being
+//! destroyed when the lifetime of the shared reference ends. The main difference is that Rust
+//! permits concurrent atomic and non-atomic reads to the same memory as those cause no issue in the
+//! C++ memory model, they are just forbidden in C++ because memory is partitioned into "atomic
+//! objects" and "non-atomic objects" (with `atomic_ref` temporarily converting a non-atomic object
+//! into an atomic object).
+//!
+//! The most important aspect of this model is that *data races* are undefined behavior. A data race
+//! is defined as conflicting non-synchronized accesses where at least one of the accesses is
+//! non-atomic. Here, accesses are *conflicting* if they affect overlapping regions of memory and at
+//! least one of them is a write. They are *non-synchronized* if neither of them *happens-before*
+//! the other, according to the happens-before order of the memory model.
 //!
-//! Each method takes an [`Ordering`] which represents the strength of
-//! the memory barrier for that operation. These orderings are the
-//! same as the [C++20 atomic orderings][1]. For more information see the [nomicon][2].
+//! The other possible cause of undefined behavior in the memory model are mixed-size accesses: Rust
+//! inherits the C++ limitation that non-synchronized conflicting atomic accesses may not partially
+//! overlap. They must be either disjoint or access the exact same memory. This in particular rules
+//! out non-synchronized differently-sized atomic accesses to the same data unless all accesses are
+//! reads.
```

Comment:
```suggestion
//! The other possible cause of undefined behavior in the memory model are mixed-size accesses: Rust
//! inherits the C++ limitation that non-synchronized conflicting atomic accesses may not partially
//! overlap. Atomic accesses must be either disjoint (non-conflicting) or access the exact same memory. This in particular rules
//! out non-synchronized differently-sized atomic accesses to the same data unless all accesses are
//! reads.
```

Talking about "They" in the second sentence carries both "non-synchronized" and "atomic" with it. "non-synchronized" and "disjoint" are mutually exclusive, and "atomic" carries significant weight for the second part of this statement: otherwise, this could be read as to admit non-synchronized writes to the exact same memory.

Referring to "Atomic accesses" as the (new) subject alleviates these concerns.

Reply:
"disjoint (non-conflicting)" makes it sound like these are equivalent terms, which they are not.

> "non-synchronized" and "disjoint" are mutually exclusive

I don't think that's true, these notions are entirely independent. It doesn't seem relevant for this paragraph, though.

Reply:
"They" here refers to "non-synchronized conflicting atomic accesses" so I think your proposal changes the meaning of this text.

Reply:
Sorry, I mixed these terms up and updated my comment.

> "disjoint (non-conflicting)" makes it sound like these are equivalent terms, which they are not.

Ok, will update the suggestion to remove that.

Reply:
However, "conflicting" and "disjoint" are mutually exclusive. So something can probably be simplified here.

Reply:
I pushed a new proposed wording.

Reply:
Right, that's in essence what I'm trying to suggest here: we can't re-use the previous subject as it carries "conflicting", so I'm trying to re-introduce "atomic accesses" in the next sentence without that adjective. And *atomic* is vital to repeat, as none of this applies to non-atomic accesses.

Reply:
The new version looks great!

Reply:
> Non-synchronized atomic accesses must be either disjoint or access the exact same memory. 

This is still wrong as they may both be reads.